### PR TITLE
magento/magento2#32309: IncosistentReferenceDefinition wrong check result

### LINKED ISSUE #32309: IncosistentReferenceDefinition wrong check result
### Preconditions (*)

Magento 2.4.1

In Magento\Framework\Setup\Declaration\Schema\Declaration\ValidationRules\ IncosistentReferenceDefinition

```
if ($column->getType() !== $referenceColumn->getType()) {
            return true;
        }
```

should be 

```
if ($column->getType() !== $referenceColumn->getType()) {
            return false;
}
``` 

Also, The class is named wrong :) (Incosistent instead of Inconsistent)

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [X] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by engcom-Foxtrot**
Re-creation of https://github.com/magento/magento2/pull/32624 as I was unable to push test coverage to original PR.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#32309

### Manual testing scenarios (*)

1. Open db_schema.xml and change type for some column with a foreign key (for example, open `app/code/Magento/User/etc/db_schema.xml`, change `user_id` column type to `float`, and remove identity attribute).
2. Run `bin/magento setup:upgrade`.
3. Upgrade should fail with message `Column definition "some_column" and reference column definition "some_other_column" are different in tables "some_table" and "some_other_table"
`

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests EE

File: dev/tests/setup-integration/testsuite/Magento/Setup/ValidationRulesTest.php [hunk 1]
```diff
@@ -87,6 +87,29 @@ public function testFailOnIncosistentReferenceDefinition()
         $this->schemaConfig->getDeclarationConfig();
     }
 
+    /**
+     * @moduleName Magento_TestSetupDeclarationModule8
+     */
+    public function testFailOnIncosistentReferenceTypeDefinition()
```
Comment (Den4ik) [inline dev/tests/setup-integration/testsuite/Magento/Setup/ValidationRulesTest.php]: Did you mean Inco**n**sistent?

**Review by Den4ik — CHANGES_REQUESTED** [review]
@engcom-Foxtrot Thanks for the PR.
Looks like I see the typo in function name.

File: dev/tests/setup-integration/testsuite/Magento/Setup/ValidationRulesTest.php (hunk 1 above)
Comment (swnsma) [inline dev/tests/setup-integration/testsuite/Magento/Setup/ValidationRulesTest.php]:
```suggestion
    public function testFailOnInconsistentReferenceTypeDefinition()
```

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-9086](https://jira.corp.magento.com/browse/ENGCOM-9086) has been created to process this Pull Request

**Review by Den4ik — APPROVED** [review]
✅ Approved

Thanks for contribution

**Comment by magento-engcom-team** [comment]
Hi @Den4ik, thank you for the review.
[ENGCOM-9086](https://jira.corp.magento.com/browse/ENGCOM-9086) has been created to process this Pull Request

**Review by swnsma — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @swnsma, thank you for the review.
[ENGCOM-9086](https://jira.corp.magento.com/browse/ENGCOM-9086) has been created to process this Pull Request

**Comment by Den4ik** [comment]
@engcom-Foxtrot Just found another one name typo. Could you please rename folder `incosistence_reference_type_definition` and change it in `ValidationRulesTest.php`

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Review by Den4ik — APPROVED** [review]
✅ Approved

**Comment by magento-engcom-team** [comment]
Hi @Den4ik, thank you for the review.
[ENGCOM-9086](https://jira.corp.magento.com/browse/ENGCOM-9086) has been created to process this Pull Request

**Comment by Den4ik** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-9086](https://jira.corp.magento.com/browse/ENGCOM-9086) has been created to process this Pull Request

**Comment by engcom-Oscar** [comment]
:heavy_check_mark: QA passed

Was checked case from Manual testing scenarios 
1. Open db_schema.xml and change type for some column with a foreign key (for example, open `app/code/Magento/User/etc/db_schema.xml`, change `user_id` column type to `float`, and remove identity attribute).
2. Run `bin/magento setup:upgrade`.
3. Upgrade should fail with message `Column definition "some_column" and reference column definition "some_other_column" are different in tables "some_table" and "some_other_table"

:x: **Before:**
Screenshot from 2021-05-17 10-28-37

:heavy_check_mark: **After:**
Screenshot from 2021-05-17 10-32-38

**Pull request merged**